### ISSUE TITLE
[FRAME Core] Add support for `view_functions`

### ISSUE THREAD

**Issue by bkchr** [paritytech/polkadot-sdk#216]
It is basically the idea from @xlc: https://forum.polkadot.network/t/wasm-view-functions/1045

I think we should start low and only add support for these view functions. For this we should add some new attribute to the FRAME macros `pallet::view_function`. It would look similar to this:

```
#[pallet::view_function]
impl<T: Config> Pallet<T> {
    pub fn account_balances(account: T::AccountId) -> Vec<T::Balance> {
         // query the storage etc
    }
}
```

This code will then generate some declarative macro `implement_view_function`:
```
macro_rules implement_view_function {
    ( $PALLET_RUNTIME_NAME:ident ) => {
        #[cfg(not(feature = "std"))]
	#[no_mangle]
	pub unsafe fn $PALLET_RUNTIME_NAME __ account_balances(input_data: *mut u8, input_len: usize) -> u64 {
		let mut #input = if input_len == 0 {
			&[0u8; 0]
		} else {
			unsafe {
				$crate::slice::from_raw_parts(input_data, input_len)
			}
		};

		let output = $PALLET_RUNTIME_NAME::account_balances(Decode::decode(input).unwrap());
		$crate::to_substrate_wasm_fn_return_value(&output)
	}
    },
}
```

This `implement_view_function` macro is then called by `construct_runtime!` to generate these functions. As we prefix each view function with the name of the pallet in the runtime there will be no conflicts for different pallet instances.

This only solves the runtime side. The node side can be implemented later, but requires more thinking.

**Comment by amiyatulu** [comment]
Instead of view function on substrate, I think offchain storing through rpc calls using for example subxt and querying from the offchain storage will be better so that we can do more computationally expensive queries. Phala network seems promising, if it can support database like [surrealdb](https://github.com/surrealdb/surrealdb/issues/213#issuecomment-1607353536).

**Comment by pgherveou** [comment]
> We could use a "hash based" index.

hashing sounds like a good idea, we just need to make sure that we take into account instantiable pallet, as there could be more than 1 occurrence of a pallet in construct_runtime.

> This is definitely good for replacing all those getter RuntimeAPIs which are mostly boilerplate. However, it looks to me as a more comfortable way to define a subset of RuntimeAPIs. I think we should take it a step further.
> 
> The design as described won't allow a contract to use those queries because they can only be called from outside the runtime and not from inside. So we probably want to design it more like a `Call`. We construct a type we might call `Getter` which is basically like `Call`. We then implement a new trait `Query` (like `Dispatchable`) on it. All the getters are then called via a single Wasm exported function which accepts `Getter`. From within the runtime we can call it directly by using `Query::query()`. This will allow a contract to pass an encoded `Query` to the runtime.
> 
> It is basically just the read-only version of `Dispatchable`.

The outer enum pallet sounds like a good idea but unlike dispatchable call we not only have to encode a call but also decode a generic Response type.

Is using `Vec<u8>` for the return type our best option here?

```rust
pub trait Query {
    fn query(&self) -> Vec<u8>
}
```

> From within the runtime we can call it directly by using Query::query()

Can't we  simply call the view function like any other fns. 
```rust
let balance = Self::account_balance(id);
```

From the perspective of an ink! contract, I imagine we would need to wrap the `Getter`  into the outer enum, and then decode the response using SCALE, something along these lines: 

```rust
let getter = Balances::Getter::account_balance{ account: id }.into();
let balance: T::Balance = self.env().query(getter).decode()?;
```

**Comment by bkchr** [comment]
> Is using `Vec<u8>` for the return type our best option here?

We probably also want to have this typed. For contracts this isn't that useful, but for UIs etc to get the type.

**Comment by pgherveou** [comment]
I mean the types are defined by the view functions inside the `#[pallet::view_functions]`, but the Query trait is  implemented by the macro and just here to (de)serialize the data, right? 

I imagine that metadata and type annotations will derive from this macro as well

**Comment by bkchr** [comment]
Ahh yeah makes sense! So we have pallet level `Query` and runtime level `RuntimeQuery`? And we implement the query trait for the runtime `RuntimeQuery`?

**Comment by pgherveou** [comment]
Yes, from what Alex wrote above, I imagine that the macro would implement the Query trait for the Pallet and Runtime, just like we have for RuntimeCall today

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/wasm-view-functions/1045/6

**Comment by ascjones** [comment]
I have been investigating this issue, at the same time Bryan @xlc has begun [designing](https://forum.polkadot.network/t/cross-consensus-query-language-xcq/7583) and developing an [`XCQ`](https://github.com/open-web3-stack/XCQ/) PoC.

There is some discussion above about whether `view_function`s as specified here would be necessary if/when `XCQ` is implemented.

> Yes, but XCM as a standard will probably never abstract over all available functionality of every available chain. And as a standard it will also never be able to evolve that fast as chains itself.

The design proposes to solve this issue with the extension model and `polkavm` programs for querying. This may or may not be viable but there is WIP to investigate this direction.

I am at a point where I have to decide where to direct my efforts:

1. Go ahead and implement `view_function` as specified here. More straightforward to implement up front, but could be superseded eventually by a more powerful `XCQ` implementation.
2. Assist with design/development of `XCQ`. More involved, larger future payoff.

At the moment I am leaning towards `2.`. What does everybody else think?

**Comment by athei** [comment]
I think `view_functions` can be thought of a special case of an XCQ program. Instead of encoding a PolkaVM program we encode a `Query` data structure similar to `Call`. I think all of the infrastructure needed for both approaches are the same. Essentially only the payload is different. Like we still need RPC's for off chain usage. And an XCM instruction for on-chain usage.

I think maybe you can implement 1. in a way that it is agnostic to the payload (Query vs. XCM query) except for the runtime.

**Comment by xlc** [comment]
XCQ will cover all use cases of view functions.

> Instead of encoding a PolkaVM program we encode a Query data structure similar to Call.

The XCQ will provide host calls to invoke extensions and one potential implementation will be calling the host call with such Query data structure so basically we can make XCQ a wrapper of view functions. 

However, that’s just one of the potential implementation. There could be other ways and each alternatives will have different trade offs. That’s something I need to figure out by analyzing the pros and cons of each potential solution and decide the best suitable one. 

So I don’t really know if we should continue do the view function as it is. It is possible the work can be directly integrated into XCQ but it is also possible the work will be obsoleted by XCQ. We need to spend more time defining XCQ spec before we can answer this question.

**Comment by athei** [comment]
> XCQ will cover all use cases of view functions.

This is understood. Just trying to find out if view_functions can be a stepping stone on the way to XCQ. Or if we should go for XCQ directly.

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/stabilizing-v16-metadata/12352/1
